### PR TITLE
Support signing requests to S3

### PR DESCRIPTION
**Pull request by rcoh**
*Issue #, if available:*  #144 
*Description of changes:* Our signer was doing a few things "wrong" that were being accepted by other services:
- not signing x-amz-date
- not signing x-amz-security token
- not signing x-amz-user-agent

There is also the issue that S3 needs the `x-amz-content-sha256` header to be set. This diff uses the new version of our signer which addresses both of those issues & adds a customization for S3 to opt into the content-256 header. The new signer deny-lists the `User-Agent` header as a special case.

This changed the signatures which a few of the test were relying on. I manually verified that the services are still happy with those signatures & updated the tests.

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: aws/rust-runtime/aws-sig-auth/src/middleware.rs [hunk 1]
```diff
@@ -91,27 +91,18 @@ impl MapRequest for SigV4SigningStage {
     type Error = SigningStageError;
 
     fn apply(&self, req: Request) -> Result<Request, Self::Error> {
-        req.augment(|req, config| {
+        req.augment(|mut req, config| {
             let (operation_config, request_config, creds) = signing_config(config)?;
 
             // A short dance is required to extract a signable body from an SdkBody, which
             // amounts to verifying that it a strict body based on a `Bytes` and not a stream.
             // Streams must be signed with a different signing mode. Separate support will be added for
```
Comment (jdisanti) [inline aws/rust-runtime/aws-sig-auth/src/middleware.rs:99]: Seems like this comment needs some tweaking with the changes below?

File: aws/rust-runtime/aws-sig-auth/src/signer.rs [hunk 2]
```diff
@@ -93,36 +97,43 @@ impl SigV4Signer {
...
+        let signable_body = request
+            .body()
+            .bytes()
+            .map(SignableBody::Bytes)
+            .unwrap_or(SignableBody::UnsignedPayload);
+        for (key, value) in aws_sigv4_poc::sign_core(request, signable_body, &sigv4_config)? {
             request
                 .headers_mut()
-                .append(key.header_name(), value.parse()?);
+                .append(HeaderName::from_static(dbg!(key)), value);
```
Comment (jdisanti) [inline aws/rust-runtime/aws-sig-auth/src/signer.rs:136]: Remove the `dbg!`?

File: aws/sdk-codegen/src/main/kotlin/software/amazon/smithy/rustsdk/SigV4SigningDecorator.kt [hunk 3]
```diff
@@ -78,20 +80,40 @@ class SigV4SigningConfig(private val sigV4Trait: SigV4Trait) : ConfigCustomizati
     }
 }
 
-class SigV4SigningFeature(private val runtimeConfig: RuntimeConfig) :
+fun needsAmzSha256(service: ServiceShape) = when {
+    service.id == ShapeId.from("com.amazonaws.s3#AmazonS3") -> true
```
Comment (jdisanti) [inline aws/sdk-codegen/src/main/kotlin/software/amazon/smithy/rustsdk/SigV4SigningDecorator.kt:84]: There's nothing in the Smithy model that tells us this?

**Review by jdisanti — COMMENTED** [review]
Mostly LGTM, just some questions and minor things.

File: aws/sdk-codegen/src/main/kotlin/software/amazon/smithy/rustsdk/SigV4SigningDecorator.kt (hunk 3 above)
Comment (rcoh) [reply aws/sdk-codegen/src/main/kotlin/software/amazon/smithy/rustsdk/SigV4SigningDecorator.kt:84]: nope signing is the wild west: https://awslabs.github.io/smithy/1.0/spec/aws/aws-auth.html#aws-auth-sigv4-trait

File: aws/rust-runtime/aws-sig-auth/src/middleware.rs (hunk 1 above)
Comment (rcoh) [reply aws/rust-runtime/aws-sig-auth/src/middleware.rs]: yep done

File: aws/rust-runtime/aws-sig-auth/src/signer.rs (hunk 2 above)
Comment (rcoh) [reply aws/rust-runtime/aws-sig-auth/src/signer.rs]: also done

**Review by jdisanti — APPROVED** [review]

**Pull request merged**